ikarus/python/dirichletvalues/dirichletvalues.hh

```diff
@@ -57,54 +61,118 @@ void registerDirichletValues(pybind11::handle scope, pybind11::class_<DirichletV
+          auto lambda    = [&](BackendType& vec, int localIndex, LocalView& lv) {
+            auto lvWrapper = LocalViewWrapper(lv);
+            function(vec.vector(), localIndex, lvWrapper);
+          };
+          self.fixBoundaryDOFs(lambda);
+
+        } else if (numParams == 4) {
+          auto& function = functor.template cast<const FixBoundaryDOFsWithIntersectionFunction>();
+          auto lambda    = [&](BackendType& vec, int localIndex, LocalView& lv, const Intersection& intersection) {
+            auto lvWrapper = LocalViewWrapper(lv);
+            function(vec.vector(), localIndex, lvWrapper, intersection);
+          };
+          self.fixBoundaryDOFs(lambda);
+
+        } else {
+          DUNE_THROW(Dune::NotImplemented, "fixBoundaryDOFs: A function with this signature is not supported");
+        }
+      },
+      pybind11::arg("functor"));
+
+  auto fixBoundaryDOFsOfSubSpaceBasis_ = [&]<typename SSB>(DirichletValues& self, const pybind11::function& functor,
+                                                           const SSB& ssb) {
+    // Disambiguate by number of arguments, as casting doesn't properly work with functions
+    pybind11::module inspect_module = pybind11::module::import("inspect");
+    pybind11::object result         = inspect_module.attr("signature")(functor).attr("parameters");
+    size_t numParams                = pybind11::len(result);
+
+    if (numParams == 2) {
+      auto& function = functor.template cast<const FixBoundaryDOFsWithGlobalIndexFunction>();
+      auto lambda    = [&](BackendType& vec, const MultiIndex& indexGlobal) {
+        // we explicitly only allow flat indices
+        function(vec.vector(), indexGlobal[0]);
+      };
+      self.fixBoundaryDOFs(lambda, ssb);
+
+    } else if (numParams == 3) {
+      auto& function = functor.template cast<const FixBoundaryDOFsWithLocalViewFunction>();
+      auto lambda    = [&](BackendType& vec, int localIndex, auto& lv) {
+        auto lvWrapper = LocalViewWrapper(lv.rootLocalView());
+        function(vec.vector(), localIndex, lvWrapper);
+      };
+      self.fixBoundaryDOFs(lambda, ssb);
+
+    } else if (numParams == 4) {
+      auto& function = functor.template cast<const FixBoundaryDOFsWithIntersectionFunction>();
+      auto lambda    = [&](BackendType& vec, int localIndex, auto& lv, const Intersection& intersection) {
+        auto lvWrapper = LocalViewWrapper(lv.rootLocalView());
+        function(vec.vector(), localIndex, lvWrapper, intersection);
+      };
+      self.fixBoundaryDOFs(lambda, ssb);
+
+    } else {
+      DUNE_THROW(Dune::NotImplemented, "fixBoundaryDOFs: A function with this signature is not supported");
+    }
+  };
+
+  // Concept for PowerBasis?
+  if constexpr (requires { Basis::PreBasis::children; }) {
+    cls.def("fixBoundaryDOFsOfSubSpaceBasis", [&](DirichletValues& self, const pybind11::function& functor, int index) {
```

Comment:
I would no propose adding this method but instead researching the bug and fixing it in dune-functions. This is cool that it works but it adds unnecessary complexity to our code.

Can you give me a small example branch which fails. Then I can look into it.

Reply:
I am not sure what bug we had there in mind, there is no real interface to fix DOFs even in C++, and not at all in Python. 
In C++ you would call `fixDOFs()` and in there you would call `forAllBoundaryDOFs(subbasis, ...)` but this method does not exist in Python. 

Thats why I added this convinience wrapper, because I think this will improve user experience with a short way of doing somthing that is often needed. Although the python code will increase in complexity. 

I can of course remove the feature

Reply:
This should by handled in python, where you add a method `fixBoundaryDOFs`, which takes the functor and a list of indices.
Then you can create a function at dirichletValues.py such that:
```python
def __fixBoundaryDOFs(dirichletValues):
    def __fixBoundaryDOFsFunc(dirichletValues,f,*args: int):
          prefixPathTypeName = "Dune::TypeTree::HybridTreePath<"
          prefixPathTypeName += ",".join("Dune::index_constant<"+str(i)+">" for i in args)
          prefixPathTypeName += ">"
  
           runCode="""
           #include <...> // header to forwardCorrectFunction
           template <typename DV>
           void fixBoundaryDofs(DV& dirichletValues, const pybind11::function& functor)
           {{
          dirichletValues.fixBoundaryDOFs(forwardCorrectFunction(functor),{prefixPathType}{{}});
           }}
           """.format(prefixPathType=prefixPathTypeName))
           return run("fixBoundaryDofs",StringIO(runCode),dirichletValues,f)
   return __fixBoundaryDOFsFunc
```

and then at `def dirichletValues(basis):` some changes at the bottom

```python
    module = generator.load(
        includes=includes, typeName=element_type, moduleName=moduleName,dynamicAttr=True  # dynamicAttr is needed to allow the addition of attributes at Python runtime to a class see https://pybind11.readthedocs.io/en/stable/classes.html#dynamic-attributes
    )
    dv= module.DirichletValues(basis)
    dv.fixBoundaryDOFs= = types.MethodType(__fixBoundaryDOFs(dv),dv) # Here we add the new function fixBoundaryDOFs
    return dv
```

Reply:
How would this look at call site?

Reply:
`dv.fixBoundaryDOFs(f,1,2,3,...)`

Reply:
so maybe This can be wrapped in syntactic sugar `dv.fixBoundaryDOFs(f,subSpaceBasisPath=(1,2,3,...))`

Reply:
Okay now thats actually really nice, I havn't yet added the syntactic suggar as you proposed, but it works already